readiness/controller/pkg/readiness/readiness_controller.go

```diff
@@ -34,14 +36,142 @@ type ReadinessReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
 func (r *ReadinessReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("readiness", req.NamespacedName)
+	readiness := &corev1alpha2.Readiness{}
+	err := r.Client.Get(ctx, req.NamespacedName, readiness)
+	if err != nil {
+		r.Log.Error(err, "Error while fetching readiness")
+		return ctrl.Result{}, err
+	}
 
-	// TODO(user): your logic here
+	updateRequired := false
+	oldStatus := readiness.Status
+
+	if len(readiness.Spec.Checks) == 0 {
+		if !oldStatus.Ready {
+			readiness.Status.Ready = true
+			readiness.Status.CheckStatus = []corev1alpha2.CheckStatus{}
+			updateRequired = true
+		}
+	} else {
+		allChecks := make(map[string][]int)
+		readiness.Status.CheckStatus = []corev1alpha2.CheckStatus{}
+		providersList := &corev1alpha2.ReadinessProviderList{}
+		err = r.Client.List(ctx, providersList)
+		if err != nil {
+			r.Log.Error(err, "Error while fetching readiness providers")
+			return ctrl.Result{}, err
+		}
+
+		for i, provider := range providersList.Items {
+			if _, ok := allChecks[provider.Spec.CheckRef]; !ok {
+				allChecks[provider.Spec.CheckRef] = make([]int, 0)
+			}
+
+			allChecks[provider.Spec.CheckRef] = append(allChecks[provider.Spec.CheckRef], i)
+		}
+
+		for _, check := range readiness.Spec.Checks {
+			checkStatus := corev1alpha2.CheckStatus{
+				Name:      check.Name,
+				Providers: make([]corev1alpha2.Provider, 0),
+				Ready:     false,
+			}
+
+			indices, ok := allChecks[check.Name]
+			if ok {
+				for _, index := range indices {
+					provider := providersList.Items[index]
+
+					if provider.Status.State == corev1alpha2.ProviderSuccessState {
+						checkStatus.Ready = true
+					}
+					checkStatus.Providers = append(checkStatus.Providers, corev1alpha2.Provider{
+						Name:     provider.Name,
+						IsActive: provider.Status.State == corev1alpha2.ProviderSuccessState,
+					})
+				}
+
+			}
+
+			readiness.Status.CheckStatus = append(readiness.Status.CheckStatus, checkStatus)
+		}
+
+		readiness.Status.Ready = true
+
+		for _, checkStatus := range readiness.Status.CheckStatus {
+			readiness.Status.Ready = readiness.Status.Ready && checkStatus.Ready
+		}
+
+		updateRequired = isUpdateRequired(oldStatus, readiness.Status)
+
+	}
+
+	if updateRequired {
+		time := metav1.Time{Time: time.Now()}
+		readiness.Status.LastComputedTime = &time
+
+		err = r.Client.Status().Update(ctx, readiness)
+		if err != nil {
+			r.Log.Error(err, "Error while updating readiness status")
+			return ctrl.Result{}, err
+		}
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func isUpdateRequired(oldStatus, newStatus corev1alpha2.ReadinessStatus) bool {
```

Comment:
Why do we need this? Even if there is no change in the status, lastComputedTime can still be changed to this iteration since `lastComputedTime` indicates the latest instance when we "refreshed" the readiness resource.

Reply:
We need to make the re-conciliation converge. Otherwise, it keeps on re-conciling till the controller process exits. So, we need this condition to converge.

Reply:
Exactly, we need this function to make the reconciliation idempotent. If we unconditionally update the `lastComputedTime` in every reconciliation cycle, that will trigger another reconiliation and it in turn will trigger another status update and so on. We need to stop updating the status if we don't have any diff between the previous status and the current computed status.

Reply:
The job controller in controller manager does similar. It checks the newly computed state and the current state to decide if an update is required. https://github.com/kubernetes/kubernetes/blob/0afa60086a0fc0b5bf6386e16732a373ad9a10b6/pkg/controller/job/job_controller.go#L871

Reply:
You can ignore status updates in your reconciliation using predicates, if that's only thing you're worried about.

Reply:
For future discussion: I don't know if `lastComputedTime` is really that useful. Instead, we could requeue the request every 1m (or pick a good number) and reconcile. That would guarantee it's not very stale.

I don't think any consumer would check the status and check if `lastComputedTime` was recent.